### ISSUE TITLE
Make the FIDO2 authenticators accesible for user sessions

### ISSUE THREAD

**Issue by swick** [systemd/systemd#25306]
### Component

other

### Is your feature request related to a problem? Please describe

Applications which want to allow users to authenticate to a service with FIDO2 currently implement some FIDO2 transports themselves. For example firefox uses authenticator-rs to use USB HID authenticators. This direct access has a few problems:

1. the app can authenticate against any service this authenticator has been registered with, in the worst case the authentication happens without any user interaction
2. apps support different transports and different software authenticators (TPM, fingerprint, etc)

The plan for solving those issues for sandboxed applications is to remove direct access to the authenticators while also providing a xdg-desktop-portal for FIDO2. This portal then provides access to the same authenticators and only with user interaction in the portal backend (the UI part of the portal).

The question is where should the authenticator implementations come from. We could use one of the libraries:

* libfido2 (used by systemd)
* authenticator-rs (firefox)
* libwebauthn (also called xdg-credentials-portal for no good reason)
* webauthn-authenticator-rs

There are also separate implementations of e.g. a TPM based authenticator (tpm-fido). In any case we would probably at some point support different authenticators than systemd or even support the same use case in different, incompatible ways.

It would be nice if there was a single source of authenticators that could be used in systemd and in xdg-desktop-portals.

### Describe the solution you'd like

These really are strawman proposals. I'd love to hear more and better ideas and if this seems feasible at all.

* systemd implements support for all kinds of authenticators and exports them via a dbus for xdg-desktop-portals to consume
* another service implements support for authenticators and exports them via dbus for systemd and xdg-desktop-portals to consume
* we choose a common library for authenticator implementations

**Comment by poettering** [comment]
what do other OSes do for this? android? chrome? windows? macos?

wouldn't it suffice for xdg-desktop-portals (or some relevant component) to proxy the HID stream from the real device after authentication? and then make the various libraries tlak to that HID proxy stream as an alternative?

**Comment by swick** [comment]
There are a bunch of different transports (HID, NFC, Bluetooth, caBLE) for roaming authenticators. System authenticators (TMP, fingerprint) don't have a transport protocol per se. That's really not the right level of abstraction we want to work with.

The client APIs for Android (https://developers.google.com/android/reference/com/google/android/gms/fido/fido2/Fido2ApiClient) and Microsoft (https://learn.microsoft.com/en-us/windows/win32/webauthn/-webauthn-portal) basically are on an abstraction level of the WebAuthn API and that's exactly what we would want to expose through a portal for apps as well.

We could implement the whole thing in xdg-desktop-portal or use a library for it but it would be even better if the system (maybe systemd) would provide something like a WebAuthn API we could work with. The only difference in this system API would be that it allows to build a flow around user interaction: when we want to get an assertion or make credentials that API should tell us that the user is supposed to, e.g. interact with the device, or expects a PIN and then we should be able to give it that pin. (obviously the API we want to provide to apps also checks if the app is allowed to authenticate to a specific relaying party and only does so when the user tells it to)

So basically what I'm asking is:

1. does systemd want to support different transports for roaming authenticators?
2. does systemd want to support platform authenticators (fingerprint, TPM)?
4. could systemd call out to another service for all the FIDO2 needs?
5. if yes to all of that: would systemd want to provide the FIDO2 service or want the service to be in a separate project?

**Comment by arianvp** [comment]
Have you seen https://github.com/AlfioEmanueleFresta/xdg-credentials-portal ?

**Comment by swick** [comment]
Yes, there are several libraries implementing different transports and authenticators. As stated, systemd only supports HID authenticators and we should figure out how to share all of that code.
